From 139a42ddf1d0d762083f1413fdcf712e22b59c1c Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Wed, 20 Dec 2023 18:58:39 +0100
Subject: [PATCH 0554/1695] drivers: stm32_remoteproc: restore firewall on low
 power exit

Save m33 access rights in rproc structure and restore
m33 access right when resuming from low power exit.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 23f6ef8a8c3878a4eea45ba4dc2ca5101c25f331 drivers: stm32_remoteproc: restore firewall on low power exit]
Signed-off-by: Gwenael Treuveur <gwenael.treuveur@foss.st.com>
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Change-Id: I048be196b7ffe1772ead3c5cab10bc9ff4233ad3
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/348410
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
---
 core/drivers/remoteproc/stm32_remoteproc.c | 46 ++++++++++++++++++----
 1 file changed, 38 insertions(+), 8 deletions(-)

diff --git a/core/drivers/remoteproc/stm32_remoteproc.c b/core/drivers/remoteproc/stm32_remoteproc.c
index 0e69585c4..7dd633e9e 100644
--- a/core/drivers/remoteproc/stm32_remoteproc.c
+++ b/core/drivers/remoteproc/stm32_remoteproc.c
@@ -7,8 +7,10 @@
 #include <config.h>
 #include <drivers/rstctrl.h>
 #include <drivers/stm32_remoteproc.h>
+#include <keep.h>
 #include <kernel/cache_helpers.h>
 #include <kernel/dt_driver.h>
+#include <kernel/pm.h>
 #include <kernel/tee_misc.h>
 #include <libfdt.h>
 #include <mm/core_memprot.h>
@@ -58,6 +60,7 @@ struct stm32_rproc_instance {
 	struct rstctrl *hold_boot;
 	paddr_t boot_addr;
 	bool tzen;
+	uint32_t m33_cr_right;
 };
 
 /**
@@ -71,6 +74,8 @@ struct stm32_rproc_instance {
 struct stm32_rproc_compat_data {
 	uint32_t rproc_id;
 	TEE_Result (*start)(struct stm32_rproc_instance *rproc);
+	TEE_Result (*pm)(enum pm_op op, unsigned int pm_hint,
+			 const struct pm_callback_handle *pm_handle);
 	bool ns_loading;
 };
 
@@ -455,20 +460,42 @@ static void stm32_rproc_cleanup(struct stm32_rproc_instance *rproc)
 	free(rproc);
 }
 
+static void stm32_rproc_a35ss_cfg(struct stm32_rproc_instance *rproc __unused)
+{
+#ifdef CFG_STM32MP25
+	stm32mp_syscfg_write(A35SSC_M33CFG_ACCESS_CR, rproc->m33_cr_right,
+			     A35SSC_M33_TZEN_CR_M33CFG_SEC |
+			     A35SSC_M33_TZEN_CR_M33CFG_PRIV);
+#endif /* CFG_STM32MP25 */
+}
+
+static TEE_Result stm32mp25_rproc_pm(enum pm_op op, unsigned int pm_hint,
+				     const struct pm_callback_handle *pm_handle)
+{
+	struct stm32_rproc_instance *rproc = pm_handle->handle;
+
+	if (PM_HINT_IS_STATE(pm_hint, CONTEXT) && op == PM_OP_RESUME)
+		stm32_rproc_a35ss_cfg(rproc);
+
+	return TEE_SUCCESS;
+}
+DECLARE_KEEP_PAGER(stm32mp25_rproc_pm);
+
 static TEE_Result stm32_rproc_probe(const void *fdt, int node,
 				    const void *comp_data)
 {
 	struct stm32_rproc_instance *rproc = NULL;
 	TEE_Result res = TEE_ERROR_GENERIC;
-#ifdef CFG_STM32MP25
-	uint32_t m33_cr_right = A35SSC_M33_TZEN_CR_M33CFG_SEC |
-				A35SSC_M33_TZEN_CR_M33CFG_PRIV;
-#endif /* CFG_STM32MP25 */
 
 	rproc = calloc(1, sizeof(*rproc));
 	if (!rproc)
 		return TEE_ERROR_OUT_OF_MEMORY;
 
+#ifdef CFG_STM32MP25
+	rproc->m33_cr_right = A35SSC_M33_TZEN_CR_M33CFG_SEC |
+			      A35SSC_M33_TZEN_CR_M33CFG_PRIV;
+#endif /* CFG_STM32MP25 */
+
 	rproc->cdata = comp_data;
 
 	if (!rproc->cdata->ns_loading) {
@@ -497,11 +524,9 @@ static TEE_Result stm32_rproc_probe(const void *fdt, int node,
 		 * Provide access rights to A35SSC_M33 registers
 		 * to the non secure context
 		 */
-		m33_cr_right = A35SSC_M33_TZEN_CR_M33CFG_PRIV;
+		rproc->m33_cr_right = A35SSC_M33_TZEN_CR_M33CFG_PRIV;
 	}
-	stm32mp_syscfg_write(A35SSC_M33CFG_ACCESS_CR, m33_cr_right,
-			     A35SSC_M33_TZEN_CR_M33CFG_SEC |
-			     A35SSC_M33_TZEN_CR_M33CFG_PRIV);
+	stm32_rproc_a35ss_cfg(rproc);
 #endif /* CFG_STM32MP25 */
 
 	/*
@@ -515,6 +540,9 @@ static TEE_Result stm32_rproc_probe(const void *fdt, int node,
 	if (!rproc->cdata->ns_loading)
 		SLIST_INSERT_HEAD(&rproc_list, rproc, link);
 
+	if (rproc->cdata->pm)
+		register_pm_driver_cb(rproc->cdata->pm, rproc, "stm32-rproc");
+
 	return TEE_SUCCESS;
 
 err:
@@ -530,12 +558,14 @@ static const struct stm32_rproc_compat_data stm32_rproc_m4_tee_compat = {
 static const struct stm32_rproc_compat_data stm32_rproc_m33_compat = {
 	.rproc_id = STM32MP2_M33_RPROC_ID,
 	.start = stm32mp2_rproc_start,
+	.pm = stm32mp25_rproc_pm,
 	.ns_loading = true,
 };
 
 static const struct stm32_rproc_compat_data stm32_rproc_m33_tee_compat = {
 	.rproc_id = STM32MP2_M33_RPROC_ID,
 	.start = stm32mp2_rproc_start,
+	.pm = stm32mp25_rproc_pm,
 	.ns_loading = false,
 };
 
-- 
2.47.2

